ms-oidc/src/main/java/net/trajano/ms/oidc/internal/OpenIdConnectResource.java

```diff
@@ -90,16 +102,31 @@ public URI authUri(@QueryParam("state") final String state,
         if (issuerConfig == null) {
             throw new BadRequestException("Invalid issuer_id");
         }
+
+        if (!issuerConfig.isOpAllowed(op)) {
+            throw OAuthTokenResponse.badRequest(ErrorCodes.INVALID_REQUEST, "Operation not allowed");
+        }
+
+        // TODO validate whether the client credentials are valid.
+
         final URI redirectUri = UriBuilder.fromUri(serviceConfiguration.getRedirectUri()).path(issuerId).build();
-        final URI buildAuthenticationRequestUri = issuerConfig.buildAuthenticationRequestUri(redirectUri, state, cryptoOps.newToken());
-        return buildAuthenticationRequestUri;
+        final String nonce = cryptoOps.newToken();
+
+        final ServerState serverState = new ServerState(state, op, nonce, authorization);
+        final String cacheKey = cryptoOps.newToken();
+
+        nonceCache.putIfAbsent(cacheKey, serverState);
+
+        return issuerConfig.buildAuthenticationRequestUri(redirectUri, cacheKey, nonce);
     }
 
     @Path("/cb/{issuer_id}")
     @GET
-    @Produces(MediaType.APPLICATION_JSON)
-    public OAuthTokenResponse callback(@QueryParam("code") final String code,
-        @PathParam("issuer_id") final String issuerId) {
+    public Response callback(@QueryParam("code") final String code,
+        @QueryParam("state") final String jwtState,
+        @PathParam("issuer_id") final String issuerId) throws IllegalArgumentException,
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a runtime exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)